### PR TITLE
Implement Initializer.__eq__

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
When sharing parameters between Blocks, gluon checks that all attributes of to-be-shared parameters match. However, as mxnet initializers don't implement `__eq__`, the check will always fail. Thus it is currently impossible to share parameters between Blocks with custom initializers.

See https://github.com/apache/incubator-mxnet/issues/16049

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Support sharing parameters between blocks with custom initializers

## Comments ##
@szha @sxjscience

### CODE REVIEW

File: python/mxnet/initializer.py [hunk 1]
```diff
@@ -268,6 +268,10 @@ def _init_default(self, name, _):
             '"weight", "bias", "gamma" (1.0), and "beta" (0.0).' \
             'Please use mx.sym.Variable(init=mx.init.*) to set initialization pattern' % name)
 
+    def __eq__(self, other):
+        if not isinstance(other, Initializer):
+            return NotImplemented
+        return type(self) == type(other) and self._kwargs == other._kwargs
```
Comment (szha) [inline python/mxnet/initializer.py]: does the second statement behave correctly if it involves constant initializer (w/ numpy array, NDArray, and mxnet numpy array)?

Comment (leezu) [reply python/mxnet/initializer.py]: We can say it behaves correctly, if we expect `Constant(10.0) != Constant(np.array([10.0])`.

Comment (leezu) [reply python/mxnet/initializer.py]: With this PR at least `Constant(10.0) == Constant(10.0)`, which was not true before...

File: tests/python/unittest/test_gluon.py [hunk 2]
```diff
@@ -3119,6 +3119,19 @@ def forward(self, x):
...
+    class MyBlock(gluon.HybridBlock):
+        def __init__(self, **kwargs):
+            super(MyBlock, self).__init__(**kwargs)
+
+            with self.name_scope():
+                self.param = self.params.get("param", shape=(1, ), init=mx.init.Constant(-10.0))
+
+    bl = MyBlock()
+    bl2= MyBlock(params=bl.collect_params())
+    assert bl.param is bl2.param
```
Comment (wkcn) [inline tests/python/unittest/test_gluon.py]: Thanks for your contribution!
Could you add the unittest 'b1.param is not bl3.param’ and the test for mx.init.Initializer?

Comment (leezu) [reply tests/python/unittest/test_gluon.py]: What do you mean? Actually `bl2 = MyBlock(params=bl.collect_params())` performs internally a check that the two initializes are equal, which will fail without this PR.

**Pull request merged**